app/components/Modal/ConfirmModal.js

```diff
@@ -91,46 +120,61 @@ export default function withModal<Props>(
         ...props
       } = this.props;
 
-      const modalOnConfirm = () => {
-        this.startWorking();
-        return onConfirm().then(result => {
-          this.stopWorking();
-          closeOnConfirm && this.state.modalVisible && this.toggleModal();
-          return result;
-        });
+      const wrapAction = (action: () => Promise<*>, closeWhenDone: boolean) => {
+        return () => {
+          const onResolve = closeWhenDone
+            ? result => {
+                this.stopWorking();
+                this.hideModal();
+                this.resetError();
+                return result;
+              }
+            : result => result;
+
+          const onError = error => {
+            this.stopWorking();
+            const errorMessage =
+              get(error, ['meta', 'errorMessage']) || 'Det skjedde en feil...';
+
+            this.setErrorMessage(errorMessage);
+            throw error;
+          };
+
+          this.startWorking();
+          return action().then(onResolve, onError);
+        };
       };
 
-      const modalOnCancel = () => {
-        this.startWorking();
-        return onCancel().then(result => {
-          this.stopWorking();
-          closeOnCancel && this.state.modalVisible && this.toggleModal();
-          return result;
-        });
-      };
+      const modalOnConfirm = wrapAction(onConfirm, closeOnConfirm);
+      const modalOnCancel = wrapAction(onCancel, closeOnCancel);
+
+      const { working, errorMessage } = this.state;
 
       return [
         <WrappedComponent key={0} {...props} onClick={this.toggleModal} />,
-        <Modal key={1} show={this.state.modalVisible} onHide={this.toggleModal}>
+        <Modal
+          closeOnBackdropClick={!working}
+          key={1}
+          show={this.state.modalVisible}
+          onHide={this.toggleModal}
+        >
           <ConfirmModal
             onCancel={modalOnCancel}
             onConfirm={modalOnConfirm}
             message={message}
             title={title}
-            disabled={this.state.working}
+            disabled={working}
+            errorMessage={errorMessage}
           />
         </Modal>
       ];
     }
   };
 }
 
-const ChildrenWithProps = ({ children, ...restProps }: { children: Node }) => (
-  <div>
-    {React.Children.map(children, child =>
-      React.cloneElement(child, { ...restProps })
-    )}
-  </div>
-);
+const ChildrenWithProps = ({ children, ...restProps }: { children: Node }) =>
+  React.Children.map(children, child =>
+    React.cloneElement(child, { ...restProps })
```

Comment:
doesn't each child need a key when you remove the div?

Reply:
Yeah, that might be true..

Reply:
> key and ref from the original element will be preserved
- Docs (https://reactjs.org/docs/react-api.html#createelement)